exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/CreateFunctionHandler.java

```diff
@@ -0,0 +1,195 @@
+    if (context.getOption(ExecConstants.DYNAMIC_UDF_SUPPORT_ENABLED).bool_val) {
+      SqlCreateFunction node = unwrap(sqlNode, SqlCreateFunction.class);
+      String jarName = ((SqlCharStringLiteral) node.getJar()).toValue();
+      String sourceName = JarUtil.getSourceName(jarName);
+
+      RemoteFunctionRegistry remoteRegistry = context.getRemoteFunctionRegistry();
+      FileSystem fs = remoteRegistry.getFs();
+      Path tmpDir = new Path(remoteRegistry.getTmpArea(), UUID.randomUUID().toString());
+      File localTmpDir = Files.createTempDir();
+
+      boolean inProgress = false;
+      try {
+        final String action = remoteRegistry.addToJars(jarName, RemoteFunctionRegistry.Action.REGISTRATION);
+        if (!(inProgress = action == null)) {
+          return DirectPlan.createDirectPlan(context, false,
+              String.format("Jar with %s name is used. Action: %s", jarName, action));
+        }
+
+        // verify that binary and source exist
+        Path remoteBinary = new Path(remoteRegistry.getStagingArea(), jarName);
+        Path remoteSource = new Path(remoteRegistry.getStagingArea(), sourceName);
+        if (!fs.exists(remoteBinary) || !fs.exists(remoteSource)) {
+          return DirectPlan.createDirectPlan(context, false,
+              String.format("Binary [%s] or source [%s] is absent in udf staging area [%s].", jarName, sourceName, remoteRegistry.getStagingArea().toUri().getPath()));
+        }
+
+        // backup binary & source (copy to udf tmp directory)
+        fs.mkdirs(tmpDir);
+        Path tmpBinary = new Path(tmpDir, jarName);
+        Path tmpSource = new Path(tmpDir, sourceName);
+
+        FileUtil.copy(fs, remoteBinary, fs, tmpBinary, false, fs.getConf());
+        FileUtil.copy(fs, remoteSource, fs, tmpSource, false, fs.getConf());
+
+        // copy binary to local fs, we don't need source for validation
+        Path localBinary = new Path(new Path(localTmpDir.toURI()), jarName);
+        fs.copyToLocalFile(tmpBinary, localBinary);
+
+        // validate functions locally
+        List<Func> functions;
+        try {
+          functions = context.getFunctionRegistry().validate(localBinary);
+        } catch (FunctionValidationException ex) {
+          return DirectPlan.createDirectPlan(context, false, ex.getMessage());
+        }
+
+        if (functions.size() == 0) {
+          return DirectPlan.createDirectPlan(context, false,
+              String.format("Jar %s does not contain functions", jarName));
+        }
+
+        // validate and register remotely
+        Jar jar = Jar.newBuilder().setName(jarName).addAllFunction(functions).build();
+        String error = register(remoteRegistry, jar, tmpBinary, tmpSource, remoteRegistry.getRetryTimes());
+
+        if (error != null) {
+          return DirectPlan.createDirectPlan(context, false, error);
+        }
+
+        // remove jars from staging area
+        try {
+          fs.delete(remoteBinary, false);
```

Comment:
Really a comment for another file, but I thought of it here...

When we register a function, how do we handle race conditions:

If we do ZK first, then the fn may not yet exist in DFS for a client that reads ZK before the move is made.

If we move the jar first, then crash, how do we clean up the "orphan" jar file?

We could use a retry to handle the first case (ZK first, then move file.)

What does this implementation use?